[WIP] Sync HTTP2 stream id between source and destination.

mike820324 opened the pull request:
This PR change the following features.

### **Slightly changed the log format** :

Put the direction(source/destination) in front of the message.

### **Upgrade Hyper-h2 to 2.5.1** :

Since hyper-h2 2.5.0, it will check whether the header field contains Host or not. As a result, update some unittests according to this.

### **Sync stream id between source and destination**:

The original code will drop some frames(especially the priority update frame. I forgot the reason why we drop these) and hence causing the stream id between source and destination unsync. This PR remove some of the code in layer/http2.py and just send the same stream id to the other endpoint.

P.S @chhsiao90 can you kindly remind me that why we added this code. Orz

### **Pending some frames**:

Due to the fact that we will pending the HEADER frame until we receive the StreamEnded event, we need to pending some of the frames that we receive until our HEADER frame is sent. 

  - PRIORITY frame
  - RST frame
  - WINDOW_UPDATE frame

### **Prevent duplicate priority frame**:

When hyper-h2 receive HEADER frame with priority updated, it will actually generate two events which are
  - RequestReceived
  - PriorityUpdated.

Based on our current implementation, 
  - we will sending out the HEADER frame with the priority updated
  - sending out PRIORITY frame with the same priority parameter of the HEADER frame.

As a result, I have added a simple logic on top of the receive_data method to check any duplicate PriorityUpdate event.

TODO:

- [x] unittest
- [ ] code cleanup

This is still WIP.  @chhsiao90 please help me review this part since you are expert in this category. Any advice is thankful.

mike820324 committed:
Console viewer enhancement.

- Adding two command line options.
- Able to dump json format.

chhsiao90 committed:
Seperate config option creation into different function for readiblity

chhsiao90 committed:
Add events_channel config option

chhsiao90 committed:
Implement event client/manager with zmq pull/push pattern

chhsiao90 committed:
Refactor MsgPublisher and proxy
1. MsgPublisher use zmq socket(no need to wrap it to zmq stream)
2. Refine code on proxy

chhsiao90 committed:
Add unit test for event manager, event client

mike820324 committed:
This commit remove the blinker dependency and modified the related
code.

chhsiao90 committed:
Refactor some domain class

chhsiao90 committed:
Implement replay

chhsiao90 committed:
Fixed req/resp to http protocol header to keep original header case

mike820324 committed:
Merge pull request #128 from mike820324/socks5-refactor

Use socks5 library for socks layer. Close Issue #128

chhsiao90 committed:
Refactor http1 layer

chhsiao90 committed:
Refine replay with using microproxy.protocol.http1 HttpConnection

mike820324 committed:
Add checking when interceptor return nothing.

mike820324 committed:
Update unittest.

mike820324 committed:
Merge pull request #133 from mike820324/remove-blinker-#67

Remove blinker dependency and modified the related code.

mike820324 committed:
Merge pull request #134 from mike820324/version-file-#130

#130 Add version file for project.

chhsiao90 committed:
Update tls layer to handle replay script

chhsiao90 committed:
Fixed some regression problem after rebase

chhsiao90 committed:
Implement tui/console run replay

mike820324 committed:
Used class member layer_manager instead of creating one

mike820324 committed:
Merge pull request #132 from mike820324/event-queue-#118

This PR add the following features.

- Add zmq push/pull into micorproxy
- Add Replay Layer 
- Clean up some of the code

chhsiao90 committed:
Resolved #139: Refactor http2 layer

mike820324 committed:
Update console viewer command line interface.
Change parameter for event client to just pass channel string

mike820324 committed:
Skip test console viewer

chhsiao90 committed:
Resolved #140: ReplayLayer handle destination tls itself

mike820324 committed:
Merge pull request #138 from mike820324/traffic-dumper-#110

Adding dump traffic file to console viewer.

chhsiao90 committed:
Implement http2 replay

chhsiao90 committed:
Resolved #141: handle h11.ConnectionClosed

chhsiao90 committed:
Handle window frame based on not exist stream_id

mike820324 committed:
Merge pull request #143 from mike820324/handle-h11-closed-#141

Resolved #141: handle h11.ConnectionClosed.

chhsiao90 committed:
Handle h2 terminate event

mike820324 committed:
Refactor the following code.
- Change LayerManager method into pure stateless functions.
- Add get_cert_store to get the instance of certificate store.

mike820324 committed:
Add get_interceptor function to get global interceptor instance.

mike820324 committed:
Will not pass proxy_server instance to EventManager.

mike820324 committed:
Refactor proxy server and EventManager:
- Will not create instance within proxy server.
- Add start_events_server in command_line.

mike820324 committed:
Update unittest.

mike820324 committed:
Merge pull request #142 from mike820324/replay-http2-#136

Implement http2 replay

chhsiao90 committed:
Enter ForwardLayer when encounter websocket upgrade

chhsiao90 committed:
Add .coveragerc

chhsiao90 committed:
Refactor and add unit test on http1 layer

mike820324 committed:
Add init_cert_store and init_interceptor functions

mike820324 committed:
Merge pull request #144 from mike820324/layer-manager-stateless

Change LayerManager into pure functions and Move EventServer out from proxy.py

chhsiao90 committed:
Enhance protocol.http1 unittest

chhsiao90 committed:
Implement protocol.http2 unittest

chhsiao90 committed:
Notify message on tui when sent replay

mike820324 committed:
Add pause resume to resolve detach problem.

chhsiao90 committed:
Update exceptions for more reasonable name and information for human

mike820324 committed:
Add MicroProxyIOStream test.

mike820324 committed:
Test MicroProxyIOStream start_tls

chhsiao90 committed:
Enhance protocol.http2 unittest

chhsiao90 committed:
Refine http2 exception

mike820324 committed:
Add test for commad line options.

mike820324 committed:
Merge pull request #149 from mike820324/tls-detach-error-#145

Resolve #145: Add pause resume to resolve detach problem.

chhsiao90 committed:
Rename xxxxxxTest to Testxxxxxx

mike820324 committed:
Omit debug layer.

mike820324 committed:
Update layer_manager and related test.

mike820324 committed:
Update layer_manager test. layer_manager related test is done.

chhsiao90 committed:
Update http1 layer from callback style into sequential style

mike820324 committed:
Merge pull request #146 from mike820324/websocket-#115

Enter ForwardLayer when encounter websocket upgrade.

chhsiao90 committed:
Fixed failed unittest on http1 for check key in headers

mike820324 committed:
Fix layer_manager run_layers reference before assign.

mike820324 committed:
Update socks layer unittest.

mike820324 committed:
Update socks error handling and unittest.

mike820324 committed:
Add urlencoded formatter.

mike820324 committed:
Use parse_qsl to parse urlencoded data.

mike820324 committed:
Merge pull request #153 from mike820324/viewer-urlencoded-#152

Add urlencoded formatter.

chhsiao90 committed:
Enhance http related unit test

chhsiao90 committed:
Enhance http related unit test

chhsiao90 committed:
First commit for http2 layer unittest

chhsiao90 committed:
Update http2 layer unit test to make sure completely read buffer from socket stream

chhsiao90 committed:
Enhance http2 unit test

chhsiao90 committed:
Enhance unittest for replay

chhsiao90 committed:
Resolved #150: upgrade gviewer to v2.1.3

mike820324 committed:
Update unittest for iostream.
- Fix incorrect exception type in test_hand_shake_faile.

chhsiao90 committed:
Add transparent proxy unittest

chhsiao90 committed:
Removed run_tests.py and update Makefile for jobs: test, coverage

mike820324 committed:
Update layer_manager unittest.
- skip some test in layer_manager when system is not linux.

mike820324 committed:
Merge pull request #148 from mike820324/unittest-#135

Refine Unittest #135.
This PR is the first part for unittest refinement.

- layer manager
- layer.socks
- layer.transparent
- layer.replay
- layer.forward
- layer.http1
- layer.http2
- protocol.http1
- protocol.http2

mike820324 committed:
Merge pull request #158 from mike820324/gviewer-upgrade-#150

Resolved #150: upgrade gviewer to v2.1.3

chhsiao90 committed:
Resolved #159: close connection after received: Connection: close from destination

chhsiao90 committed:
Enhance socks layer error handling

chhsiao90 committed:
Try format xml/json while handling uriformatted format request

chhsiao90 committed:
Merge pull request #160 from mike820324/http1-redirect-#159

Resolved #159: close connection after received: Connection: close

mike820324 committed:
Add options for enable/disable destination certificate verification.

mike820324 committed:
Add command line utility to

- create server certificate.
- download trusted ca from mozilla.

mike820324 committed:
Merge pull request #161 from mike820324/refine-stream-closed-exec-#157

Refine exception handling about StreamClosedError

mike820324 committed:
Exception handling when certificate verification failed.

mike820324 committed:
Merge pull request #163 from mike820324/urlencoded-format-#162

Try format xml/json while handling uriformatted format request

mike820324 committed:
Use certifi to check trusted pem location.
Change default insecure to no.

mike820324 committed:
Bump pyzmq version to 15.4.0

mike820324 committed:
Merge pull request #165 from mike820324/insecure-cert-#164

Resolve#164:  Adding options to enable/disable verification of insecure cert.

mike820324 committed:
Fix urlencoded issue.

mike820324 committed:
Update based on review.

mike820324 committed:
Update iostream and related unitttest.

- MicroProxySSLIOstream will now handle connect.
- Modify unittest to test all iostream.

mike820324 committed:
Merge pull request #167 from mike820324/fix-urlencoded-#166

Resolve #166: Fix urlencoded issue.

mike820324 committed:
Fix unittest in linux.

mike820324 committed:
Add ZeroReturnError handling in iostream.

mike820324 committed:
Code cleanup and add some comment.

- Move socket.error to SSL.SysCallError
- Remove redundant code.
- Add some side note for MicroProxySSLIOStream.

mike820324 committed:
Merge pull request #168 from mike820324/iostream-refactor

Update iostream and related unitttest.

mike820324 committed:
Add command line options to specify the max width of summary view.

Default to 80.

mike820324 committed:
Fix indent error.

mike820324 committed:
Use get_terminal_size to determine the terminal width

mike820324 committed:
Merge pull request #170 from mike820324/ui-width-#151

Resolved #151: Change the max terminal width size when start up.

chhsiao90 committed:
Upgrade gviewer to v3.0.1

chhsiao90 committed:
tui with pygments color scheme

mike820324 committed:
Update gitignore

chhsiao90 committed:
Fixed error handling on http1 failed problem

mike820324 committed:
Remove redundant code from main.py

mike820324 committed:
Merge pull request #172 from mike820324/gviewer-upgrade-#169

Gviewer upgrade #169

chhsiao90 committed:
Resolved #155: add exported indicator

chhsiao90 committed:
Implement log publisher

chhsiao90 committed:
Resolved #174: handle connection closed on response without Content-Length, chunked defined

mike820324 committed:
Move create zmq socket logic from event.manager to utils.

mike820324 committed:
Remove zmq creation logic from msg_publisher.

mike820324 committed:
Add init_system_logger and change get_logger parameters.

- explicit called init_system_logger instead of init logger when load
  module.

- explicit pass root logger instance to register_lob_publiser.

mike820324 committed:
Add certstore unittest.

mike820324 committed:
Merge pull request #177 from mike820324/cert-store-unittest

Add certstore related unittest.

mike820324 committed:
Merge pull request #181 from mike820324/http-resp-close-#174

Resolved #174: handle connection closed on response

mike820324 committed:
Merge pull request #178 from mike820324/exported-indicator-#155

Resolved #155: add exported indicator

chhsiao90 committed:
Resolved #173: decode base64 when send replay request

chhsiao90 committed:
Upgrade gviewer to v3.0.3

chhsiao90 committed:
Add replay http2 post body unittest

mike820324 committed:
Merge pull request #182 from mike820324/http1-replay-failed-#173

Resolved #173: decode base64 when send replay request

mike820324 committed:
Merge pull request #180 from mike820324/publish-log-#116

Resolved #116: Add ability to publish log message to zmq.

chhsiao90 committed:
Refactor layer.test_http1 unittest

chhsiao90 committed:
Enhance layer.test_http1 unittest

chhsiao90 committed:
Refactor socks layer and enhance unittet

mike820324 committed:
Merge pull request #183 from mike820324/http1-test-refactor

Http1 test refactor

chhsiao90 committed:
Move viewer/format/*.py into viewer/ to prevent duplicate with built-in function: format

chhsiao90 committed:
Update console listener channel with topic: message, related to issue #116

chhsiao90 committed:
Deserialize to ViewerContext on viewer

chhsiao90 committed:
Implement pygments highlighting

chhsiao90 committed:
Add defulat config for console-viewer

chhsiao90 committed:
Add pygments dependency

mike820324 committed:
Merge pull request #184 from mike820324/socks-refactor

Refactor socks layer and enhance unittet

chhsiao90 committed:
Upgrade gviewer v3.0.4

chhsiao90 committed:
Refactor config: Remove Config and enhance unittest

mike820324 committed:
Merge pull request #188 from mike820324/config-refactor

Refactor config: Remove Config and enhance unittest

chhsiao90 committed:
Refactor TlsLayer.

Abstract the TlsLayer logic into two part.
So that it is more readable and testable.

- layer
- protocol.

mike820324 committed:
Merge pull request #187 from mike820324/pygments-#185

Pygments #185

chhsiao90 committed:
Refactor iostream: though the variable: socket is duplicate with module name: socket, rename it

chhsiao90 committed:
Enrich our StreamClosedError with which type of it

chhsiao90 committed:
Add unittest on TLS related code.

- protocol.tls
- layer.tls

chhsiao90 committed:
Upgrade gviewer to v3.0.5

mike820324 committed:
Remove extra function defines.

When verify mode is SSL.VERIFY_NONE,
the SSL.ERROR exception will not be raised whatever the return value is.

chhsiao90 committed:
Implement formatter unittest

mike820324 committed:
Remove verify hostname logic in start_tls in tls protocol.

Since the verification logic is already within MicroProxySSLIOStream,
I remove the verification logic in start_tls in tls protocol.

Moreover, I also modify the MicroProxySSLIOStream to properly raise the
VerificationError when hostname is incorrect.

chhsiao90 committed:
Upgrade gviewer to v3.0.6

chhsiao90 committed:
Merge pull request #194 from mike820324/gviewer-upgrade-3.0.6

Upgrade gviewer to v3.0.6

mike820324 committed:
Merge pull request #189 from mike820324/tls-test

Refactor TLS related code and add unittest.

mike820324 committed:
Merge pull request #195 from mike820324/formatter-test-#192

Implement formatter unittest.

mike820324 committed:
Organize the tornado extened classes.

- Create folder called tornado_ext
- Move iostream.py to tornado_ext
- Separate some of the logic from ProxyServer to tornado_ext.
- Update related clients and unittest.

mike820324 committed:
Merge pull request #196 from mike820324/tornado-ext

Organize the tornado extended classes.

mike820324 committed:
Fix bug when moving ProxyServer to tornado_ext

mike820324 committed:
Add server state related context and functions.

- Add ServerContext.
- Add init_server_state.
- Move initialize function to init_server_state.
- Modify LayerContext.

mike820324 committed:
Apply clients to used the new server state context and function.

mike820324 committed:
Fix unittest.

mike820324 committed:
Add unitetest for server state related code.

mike820324 committed:
Merge pull request #197 from mike820324/server-state-#147

 Resolved #147: Add Server state and related API.

mike820324 committed:
Remove pause/resume methods.

- When the iostream is not in reading state,
  we will not handle_read even when event is happened.

mike820324 committed:
Move layer_manger to layer/manager.py.

mike820324 committed:
Add peek method and porting some mitmproxy code.

- Implement peek method for tornado iostream.
- Use peek method to get alpn and sni.
- Destination stream can now start_tls in non-blocking mode.

mike820324 committed:
Update unittest.

- Move some shared logic into test/utils.py
- Fix test error when refactor.

mike820324 committed:
Add missing __init__.py in test/tornado_ext.

mike820324 committed:
Merge pull request #199 from mike820324/remove-pause-resume

Remove pause/resume methods from microproxy iostream.

mike820324 committed:
Merge pull request #198 from mike820324/peeking-tls

Rewrite TLS related code.

mike820324 committed:
Refine config.py and command line interface.

- Add new option types.
- Properly handle boolean type.
- Change subcommand into individual binary.

mike820324 committed:
Update unittest.

mike820324 committed:
Update setup.py and Makefile.

- Can use pip install to install the different binary interface.
- Add some shortcut in Makefile to install mpxxx in virtualenv.

mike820324 committed:
Merge pull request #201 from mike820324/refine-config-#137

Resolve #137 Refine Config and command line interface.

mike820324 committed:
Update README and CHANGELOG.

mike820324 committed:
bump version to 0.4.0

mike820324 committed:
Upgrade h11 to 0.6.0-dev.

Since the reason support of h11 is still in master branch,
we need to install h11 from github.

mike820324 committed:
Bump version to 0.4.0+dev and update readme.

mike820324 committed:
Merge pull request #206 from mike820324/h11-0.6.0-#205

Upgrade h11 to 0.6.0-dev.

chhsiao90 committed:
Add tls in ViewerContext

chhsiao90 committed:
Record TLS info and display it in tui

chhsiao90 committed:
Pass the socks layer unit test

mike820324 committed:
Add socks5 v0.1.0 tag point in setup.py and proxy.txt.

mike820324 committed:
Merge pull request #210 from mike820324/socks-update-#209

Pass the socks layer unit test

mike820324 committed:
Bump version to 0.4.1 due to critical issue.

mike820324 committed:
Bump version to v0.4.1+dev.

mike820324 committed:
Fix incorrect pip options in Makefile.

chhsiao90 committed:
Update requirements/proxy.txt h11 to njsmith/h11

mike820324 committed:
Merge pull request #208 from mike820324/tls-info-#204

Resolved #204: Display TLS Information

chhsiao90 committed:
unit test on ReplayHandler will use its io_loop that will let io_loop for test behavior weird

chhsiao90 committed:
Refactor event related code

chhsiao90 committed:
Add version in ViewerContext

chhsiao90 committed:
Fixed some bug when handling replay event

mike820324 committed:
Add default value for some of the options in mpdump.

mike820324 committed:
Merge pull request #214 from mike820324/h11-requirements

Update requirements/proxy.txt h11 to njsmith/h11

mike820324 committed:
Merge pull request #216 from mike820324/fixed-failed-test-#211

Fixed unit test failed on OSX problem

mike820324 committed:
Merge pull request #219 from mike820324/mpdump-default-option-#218

Resolved #218: add default value for mpdump.

mike820324 committed:
Add __neq__ to coverage ignore list.

mike820324 committed:
Merge pull request #217 from mike820324/ctx-version-#203

Add version in ViewerContext

chhsiao90 committed:
Refactor context deserialization

mike820324 committed:
Use commit hash to tag h11 version instead of following the master branch.

To avoid the potential api change when h11 master change merged new commits.

chhsiao90 committed:
Resolved #193: add tui unit test

mike820324 committed:
Merge pull request #223 from mike820324/tag-h11-commit-hash

Use commit hash to tag h11 version.

mike820324 committed:
Merge pull request #222 from mike820324/refine-deserialize-ctx

Refactor context deserialization

chhsiao90 committed:
Disabled tornado log on test

mike820324 committed:
Merge pull request #225 from mike820324/tui-test-#193

Resolved #193: add tui unit test

mike820324 committed:
Merge pull request #226 from mike820324/disable-tornado-log-on-test

Disabled tornado log on test

chhsiao90 committed:
Add missing unittest

mike820324 committed:
Merge pull request #227 from mike820324/unittest

Add missing unittest

chhsiao90 committed:
Refine the layers code
1. Extract Layer as an interface, and extends ApplicationLayer, ProxyLayer with it
2. add shortcut for src_stream, dest_stream, interceptor, config

mike820324 committed:
Merge pull request #228 from mike820324/layer-refine

Refine the layers code.

mike820324 committed:
Fix proxy error when connecting to https website.

mike820324 committed:
Update Makefile.

- when run make test, the pyc will not be created.

mike820324 committed:
Cleanup development files.

- remove main.py
- move application.cfg and log.cfg to conf folder.

mike820324 committed:
Merge pull request #230 from mike820324/fix-tls-failed

Fix proxy error when connecting to https.

mike820324 committed:
Refactor proxy logging initialization mechanism.

- Add a static class ProxyLogger.
- Apply to all the file using logger
- Add three command line options
  - log-level
  - log-file
  - log-config

mike820324 committed:
Merge pull request #229 from mike820324/logging-refactor-#175

Refactor the logger framework to accept more configurations.

mike820324 committed:
Update unittest.

mike820324 committed:
Fix proxy start failed when using logger config file.

chhsiao90 committed:
Implement http proxy

chhsiao90 committed:
Fixed infinte loop problem that layer manager enter http1 layer after exit http1 layer

chhsiao90 committed:
Add some unit test

mike820324 committed:
Upgrade h11 to 0.7.0.

mike820324 committed:
Upgrade h2 version to 2.4.2.

The 2.5.0 has some api change which may need some code modification.

chhsiao90 committed:
Handle proxy path with more reasonable ways

mike820324 committed:
Add new entry in Makefile.

Add run-server entry in Makefile, which will point to the main config
location.

mike820324 committed:
Logger will use full name instead of shortname.

- modify log.py
- modify logging.cfg

mike820324 committed:
Merge pull request #231 from mike820324/dev-file-cleanup

Cleanup development files.

mike820324 committed:
Merge pull request #233 from mike820324/logger-quickfix

Fix proxy start failed when using logger config file.

mike820324 committed:
Merge pull request #236 from mike820324/h11-upgrade

Upgrade h11 to 0.7.0 and hyper-h2 to 2.4.2.

mike820324 committed:
Add log-config-file options to Makefile run-server.

mike820324 committed:
Merge pull request #235 from mike820324/http-proxy

Implement http proxy

mike820324 committed:
Upgrade construct to 2.8.8 and modify pyca_tls

mike820324 committed:
Upgrade socks5 to 0.2.1.

In this commit, all the require dependency links can be removed.

mike820324 committed:
Remove process-dependency-links from Makefile.

mike820324 committed:
Add unittest for TlsClientHello.

chhsiao90 committed:
Mark IOStream as closed while it had been detached, so that it wont raise Exception when try close it

mike820324 committed:
Fix based on review.

Use explicit type when testing.

mike820324 committed:
Merge pull request #237 from mike820324/socks-upgrade

Upgrade socks5 to v 0.2.1

mike820324 committed:
Merge pull request #238 from mike820324/error-handling-#234

Mark closed over detached IOStream

chhsiao90 committed:
Disabled file logger by default and set up default log level with INFO

chhsiao90 committed:
Add some default config pathes

chhsiao90 committed:
Add mpserver.log in .gitignore

mike820324 committed:
Merge pull request #240 from mike820324/default-config-path

Add some default config path.

chhsiao90 committed:
Enhance debug log message

mike820324 committed:
Merge pull request #242 from mike820324/enhance-debug-log-msg

Enhance debug log message

mike820324 committed:
Change http2 log format a bit.

mike820324 committed:
Add protocol_http2 entry in conf/logging.cfg.

mike820324 committed:
Sync the stream id between client and server.

- Remove some of the code from application/http2.py.

mike820324 committed:
Pending events when we receive header frame but not yet sending out yet.

mike820324 committed:
Prevent from sending duplicate PRIORTY frame when RequestReceived event has priority updated.

chhsiao90 commented:
Wow, finally, you remembered this repository!

chhsiao90 commented:
Simply ask, you drop the src <-> dest stream id mapping, and just straight forward with same stream id ?
I think that would be ok!

mike820324 commented:
Yes, I have drop the steam id mapping. Do you remember why we add this mapping code, I believe there used to be a bug related to this, but I forgot. :(

chhsiao90 commented:
I remembered that legacy hyper-h2 didn't support for jumping the stream id.

mike820324 commented:
Oh, thanks for reminder. So it should be okay to remove the mappings part of code?

chhsiao90 commented:
If the test is ok, I think it's fine to remove!

mike820324 committed:
Upgrade to hyper-h2 2.5.1.

Hyper h2 now force each request must contains a Host field.
As a result, fix some unit tests by adding Host field.

mike820324 committed:
Fix unittest failed.

chhsiao90 commented on conf/logging.cfg:
```diff
@@ -1,5 +1,5 @@
 [handlers]
-keys=consoleHandler
+keys=consoleHandler, fileHandler
```
I think fileHandler by default sometimes inconvenient...
We don't see the file for debug or troubleshoot...
Is there any reason for fileHandler by default?

mike820324 replied:
My bad, I misinterpret the handlers sections. I thought it will just define a handler and won't use it until some logger use the handler.

I'll remove this part. Nice catch.

chhsiao90 commented on microproxy/protocol/http2.py:
```diff
@@ -38,6 +39,8 @@ def decorate(*args):
         self.stream = stream
 
+        self.on_pending_events = on_pending_events or on_unhandled_with_type("PENDING_EVENTS")
+        self.on_header_receive = on_header_receive or on_unhandled_with_type("REQUEST_HEADER_RECEIVE")
```
If you have this event, then I think we SHOULD NOT have request event and response, but with header event and data event.

Because the header event will be sent twice with on_header_receive callback and on_request or on_response callback...

mike820324 replied:
The ```on_header_receive``` will not actually sending the header event. Instead, what it will do is just marking the stream as a **pending event stream**.  (I should probably change the ```self.on_header_receive``` to other much proper name. )
So that any other event received after that will be put into the **self.pending_events_of_stream**, and will be send to destination after we have receive the **EndStream** event.

chhsiao90 replied:
Not quite straightforward for me.
Could you describe how this two events work in our Http2Layer?
Is that there is any unhandled circumstance that need these two event?
Thanks!

chhsiao90 reviewed (approved):
Overall it's looks great!
Had just some comment for you.

chhsiao90 commented on microproxy/protocol/http2.py:
```diff
@@ -166,26 +183,28 @@ def handle_update_settings(self, event):
 
     def handle_window_updates(self, event):
-        self.on_window_updates(event.stream_id, event.delta)
+        self.on_window_updates(event.stream_id, event.delta, event)
```
If all callback send event as parameter, then I think there is no need to send the field of the event.
It's duplicate!

chhsiao90 commented:
That me simply describe what I understand about the pending event and header receive event.

You introduce pending event is for the purpose to delay the event sending to the destination stream of  WINDOW_UPDATE, PRIORITY, RESET that the stream still in idle state.
And the header receive is just fulfill the pending event functionality.

But why we need this? Is the hyper-h2 restriction?
